### ISSUE TITLE
1.7.26 regression: NoMethodError: private method `require' called for Kernel:Module in Sequel

### ISSUE THREAD

**Issue by perlun** [jruby/jruby#4155]
Something has gone broken in JRuby 1.7.26, causing the error below.

```shell
$ bundle exec irb
The signal IOT is in use by the JVM and will not work correctly on this platform
The signal CLD is in use by the JVM and will not work correctly on this platform
The signal EXIT is in use by the JVM and will not work correctly on this platform
jruby-1.7.26 :001 > require 'sequel'
 => true
jruby-1.7.26 :002 > Sequel.extension :core_extensions
NoMethodError: private method `require' called for Kernel:Module
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/sequel-4.38.0/lib/sequel/core.rb:139:in `extension'
	from org/jruby/RubyArray.java:1613:in `each'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/sequel-4.38.0/lib/sequel/core.rb:139:in `extension'
	from (irb):2:in `evaluate'
	from org/jruby/RubyKernel.java:1079:in `eval'
	from org/jruby/RubyKernel.java:1479:in `loop'
	from org/jruby/RubyKernel.java:1242:in `catch'
	from org/jruby/RubyKernel.java:1242:in `catch'
	from /Users/plundberg/.rvm/rubies/jruby-1.7.26/bin/irb:13:in `(root)'
	from org/jruby/RubyKernel.java:1059:in `load'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli/exec.rb:1:in `(root)'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli/exec.rb:27:in `run'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli.rb:332:in `exec'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli.rb:20:in `dispatch'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/cli.rb:11:in `start'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/exe/bundle:34:in `(root)'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/lib/bundler/friendly_errors.rb:100:in `with_friendly_errors'
	from org/jruby/RubyKernel.java:1059:in `load'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/gems/bundler-1.13.1/exe/bundle:26:in `(root)'
	from org/jruby/RubyKernel.java:1079:in `eval'
	from /Users/plundberg/.rvm/gems/jruby-1.7.26/bin/bundle:1:in `(root)'jruby-1.7.26 :003 > quit
```

When I switch to JRuby 1.7.25, all is fine:

```
$ rvm use 1.7.25
Using /Users/plundberg/.rvm/gems/jruby-1.7.25
$ bundle exec irb
jruby-1.7.25 :001 > require 'sequel'
 => true
jruby-1.7.25 :002 > Sequel.extension :core_extensions
 => [:core_extensions]
jruby-1.7.25 :003 > quit
```

The failing code (https://github.com/jeremyevans/sequel/blob/master/lib/sequel/core.rb#L139) looks like this:

```ruby
  def self.extension(*extensions)
    extensions.each{|e| Kernel.require "sequel/extensions/#{e}"}
end
```

Any ideas or suggestions on what change in 1.7.26 could have been causing this?

/cc @olleolleolle, @sdahlbac

**Comment by olleolleolle** [comment]
A workaround:

```ruby
require 'sequel'
module Sequel
  def self.extension(*extensions)
    extensions.each { |e| Kernel.send(:require, "sequel/extensions/#{e}") } # PATCH: avoid Kernel.require being private
  end
end
Sequel.extension :core_extensions
```

**Comment by jayjlawrence** [comment]
There is a similar issue with Bundler 1.13 and 1.13.1. The issue is also present in jRuby 1.7.24. I haven't checked other versions. See the referenced issue above.

**Comment by olleolleolle** [comment]
@perlun Perhaps your jruby-1.7.25 has bundler v1.12.5 installed?

**Comment by jayjlawrence** [comment]
Stack trace shows bundler 1.13.1 - this may be a variation on the bundler problem.

@perlun try 'gem uninstall bundler' and if given the choice remove the 1.13.x versions. Then 'gem install bundler -v=1.12.5' and  see if this clears up the issue.

**Comment by headius** [comment]
This is very strange. Kernel.require appears to be public and callable in all recent JRuby versions I tried.

**Comment by headius** [comment]
Ok, so the problem is definitely from `#reverse_rubygems_kernel_mixin', which attempts to restore the original `require` over the top of the patched version from RubyGems. In this case, it yanks `gem_original_require` out as an unbound method and redefines it as `require`. It basically transplants the old one back in.

https://github.com/bundler/bundler/commit/c144eb67490d11db85a808661402a54ee297ba0c#diff-f271319d553ee0af23e993e254a92c6cR293

In MRI, that causes visibility to be set to whatever is current for the frame, which is usually public. So the new `require` is public.

In JRuby, because we associate visibility with the method rather than with the method table entry, the original `private` visibility sticks to the method object and it remains private after transplanted. So this logic causes `require` to be private when it was public.

Working with bundler folks on a fix now in bundler/bundler#4975

**Comment by headius** [comment]
We have proposed to bundler that they modify the `redefine_method` logic to also `send(:public, ...`) after the transplanting. It will hurt no Ruby and will allow the logic to work on any JRuby version.

FWIW, recent 9k (since 9.1.3.0) does not appear to have this problem. I fixed #3869 in 9b4e97b283c92967f74806cf4529a2405112f163 on master but did not backport that fix to 1.7. All versions of JRuby prior to 1.7.27 and 9.1.3.0 will have this bug.
